File: JAVA/InterleavingString_recursive.java
class Solution {
    public boolean isInterleave(String s1, String s2, String s3) {
        // Start typing your Java solution below
        // DO NOT write main() function
//        if(s1 == null || s2 == null || s3 == null)
//        	return false;
        if(s1.length() == 0 && s2.length() == 0 && s3.length() == 0)
        	return true;
//        if(s1.equals("") || s2.equals("") || s3.equals(""))
//        	return false;
        if(s3.length() == 0)
        	return false;
        char ch = s3.charAt(0);
        if(s1.length() == 0)
        {
        	if(s2.length() == 0)
        		return false;
        	if(s2.charAt(0) == ch)
        		return isInterleave(s1, s2.substring(1), s3.substring(1));
        	else
        		return false;
        }
        else
        {
        	if(s1.charAt(0) == ch)
        	{
        		boolean tag = isInterleave(s1.substring(1), s2, s3.substring(1));
        		if(tag)
        			return true;
        	}
        	if(s2.length() != 0)
        	{
        		if(s2.charAt(0) == ch)
        			return isInterleave(s1, s2.substring(1), s3.substring(1));
        	}
        	return false;
        }
    }
}
public class InterleavingString_recursive {

	/**
	 * @param args
	 */
	public static void main(String[] args) {
		// TODO Auto-generated method stub

	}

}

File: C++/InterleavingString_recursive.cpp
#include<iostream>
#include<vector>
#include<queue>
#include<string>
#include<string.h>
#include<math.h>
#include<algorithm>
using namespace std;
class Solution {
public:
	bool solve(string& s1, int index1, int len1, string& s2, int index2, int len2, string& s3, int index3, int len3)
	{
		if(index1 == len1 && index2 == len2)
		{
			return index3 == len3;
		}
		else if(index1 == len1)
		{
			return s2.substr(index2) == s3.substr(index3);
		}
		else if(index2 == len2)
			return s1.substr(index1) == s3.substr(index3);
		if(s1[index1] != s3[index3] && s2[index2] != s3[index3])
			return false;
		else 
		{
			if(s1[index1] == s3[index3] && solve(s1, index1+1, len1, s2, index2, len2, s3, index3+1, len3))
				return true;
			else
				return solve(s1, index1, len1, s2, index2+1, len2, s3, index3+1, len3);
		}
	}
    bool isInterleave(string s1, string s2, string s3) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
		int len1 = s1.size();
		int len2 = s2.size();
		int len3 = s3.size();
		if(len1 + len2 != len3)
			return false;
		return solve(s1, 0, len1, s2, 0, len2, s3, 0, len3);
    }
};